project/src/main/control-manager.gd

```diff
@@ -0,0 +1,131 @@
+"""
+func input_event_from_json(json: Dictionary) -> InputEvent:
+	if not json:
+		return null
+	
+	var input_event: InputEvent
+	match json.get("type"):
+		"key":
+			var input_event_key := InputEventKey.new()
+			input_event_key.scancode = json.get("scancode")
+			input_event = input_event_key
+		"joypad_button":
+			var input_event_joypad_button := InputEventJoypadButton.new()
+			input_event_joypad_button.device = json.get("device")
+			input_event_joypad_button.button_index = json.get("button_index")
+			input_event = input_event_joypad_button
+		_:
+			push_warning("Unrecognized input event type: '%s'" % [json.get("type")])
+	return input_event
+
+
+"""
+Converts an InputEvent to a json dictionary.
+
+This supports InputEventKey and InputEventJoypadButton events. It does not support joystick, mouse or touch events.
+"""
+func input_event_to_json(input_event: InputEvent) -> Dictionary:
+	var json := {}
+	if input_event is InputEventKey:
+		json["type"] = "key"
+		json["scancode"] = input_event.scancode
+	elif input_event is InputEventJoypadButton:
+		json["type"] = "joypad_button"
+		json["device"] = input_event.device
+		json["button_index"] = input_event.button_index
+	return json
+
+
+"""
+Returns a string representation of an input, suitable for showing to the player.
+
+Parameters:
+	'input_json': A json representation of a keyboard or joypad input
+
+Returns:
+	A short human-readable string like 'DPAD Left' or 'Escape'
+"""
+func pretty_string(input_json: Dictionary) -> String:
+	var result: String
+	match input_json["type"]:
+		"key":
+			result = OS.get_scancode_string(input_json["scancode"])
+		"joypad_button":
+			result = Input.get_joy_button_string(input_json["button_index"])
+			result = result.replace("Face Button", "Face")
+	return result
+
+
+"""
+Updates the InputMap with the bindings in the specified json file
+"""
+func _bind_keys_from_file(path: String) -> void:
+	var json_text := FileUtils.get_file_as_text(path)
+	var json_dict: Dictionary = parse_json(json_text)
+	_bind_keys_from_json_dict(json_dict)
+
+
+"""
+Updates the InputMap with the bindings in the specified json
+"""
+func _bind_keys_from_json_dict(json: Dictionary) -> void:
+	for action_name in json:
+		var input_events := []
+		for input_event_json in json.get(action_name):
+			var input_event := input_event_from_json(input_event_json)
+			if input_event:
+				input_events.append(input_event)
+		
+		_bind_keys(action_name, input_events)
+
+
+"""
+Updates a single InputMap action with the specified bindings
+
+This replaces all bindings for the specified action with a new set of bindings.
+
+Parameters:
+	'action_name': the action to rebind
+	
+	'input_events': the InputEvents to bind
+"""
+func _bind_keys(action_name: String, input_events: Array) -> void:
+	# add this keybind in case it doesn't exist
+	if !InputMap.has_action(action_name):
+		InputMap.add_action(action_name)
+	
+	# erase any already bound events from this action_name
```

Comment:
This comment should probably refer to `action` and not `action_name`

Reply:
Actually, this function's comments seem pretty unnecessary. The code speaks for itself.

Reply:
I've removed these comments.